powerauth-java-crypto/src/test/java/io/getlime/security/powerauth/crypto/lib/totp/TOTPTest.java

```diff
@@ -0,0 +1,110 @@
+    void testValidateTOTPCurrentStep(final long seconds, final @ConvertWith(DateTimeConverter.class) LocalDateTime localDateTime, final String step, final String otp, final String algorithm) throws Exception {
+        final boolean result = switch (algorithm) {
+            case "HmacSHA1" -> TOTP.validateTOTPSHA1(otp, SEED, localDateTime);
+            case "HmacSHA256" -> TOTP.validateTOTPSHA256(otp, SEED, localDateTime);
+            case "HmacSHA512" -> TOTP.validateTOTPSHA512(otp, SEED, localDateTime);
+            default -> throw new AssertionFailedError("Not supported algorithm " + algorithm);
+        };
+        assertTrue(result);
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/io/getlime/security/powerauth/crypto/lib/totp/data.csv", useHeadersInDisplayName=true)
+    void testValidateTOTPOneStepBack(final long seconds, final @ConvertWith(DateTimeConverter.class) LocalDateTime localDateTime, final String step, final String otp, final String algorithm) throws Exception {
+        final LocalDateTime movedLocalDateTime = localDateTime.plusSeconds(30);
+
+        final boolean result = switch (algorithm) {
+            case "HmacSHA1" -> TOTP.validateTOTPSHA1(otp, SEED, movedLocalDateTime);
+            case "HmacSHA256" -> TOTP.validateTOTPSHA256(otp, SEED, movedLocalDateTime);
+            case "HmacSHA512" -> TOTP.validateTOTPSHA512(otp, SEED, movedLocalDateTime);
+            default -> throw new AssertionFailedError("Not supported algorithm " + algorithm);
+        };
+        assertTrue(result);
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/io/getlime/security/powerauth/crypto/lib/totp/data.csv", useHeadersInDisplayName=true)
+    void testValidateTOTPTwoStepsBack(final long seconds, final @ConvertWith(DateTimeConverter.class) LocalDateTime localDateTime, final String step, final String otp, final String algorithm) throws Exception {
+        final LocalDateTime movedLocalDateTime = localDateTime.plusSeconds(60);
+
+        final boolean result = switch (algorithm) {
+            case "HmacSHA1" -> TOTP.validateTOTPSHA1(otp, SEED, movedLocalDateTime);
+            case "HmacSHA256" -> TOTP.validateTOTPSHA256(otp, SEED, movedLocalDateTime);
+            case "HmacSHA512" -> TOTP.validateTOTPSHA512(otp, SEED, movedLocalDateTime);
+            default -> throw new AssertionFailedError("Not supported algorithm " + algorithm);
+        };
+        assertFalse(result);
+    }
+
+    static class DateTimeConverter extends TypedArgumentConverter<String, LocalDateTime> {
+        protected DateTimeConverter() {
+            super(String.class, LocalDateTime.class);
+        }
+
+        @Override
+        protected LocalDateTime convert(final String source) throws ArgumentConversionException {
+            try {
+                return LocalDateTime.parse(source);
+            } catch (Exception e) {
```

Comment:
Can we add a debug log for this exception before converting it?

Reply:
I believe that it is not needed. This is a test converter and it is clearly propagated to the test report.